### DISCUSSION TITLE
I2C garbage data on the Pi Pico VS RPi4

### DISCUSSION THREAD

**Discussion by ajevticc** [micropython/micropython#11671] [RP2040 / Pico] [answered]
Hello everyone!

So, I am interfacing a temperature sensor MCP9600. The data read out works just fine.  

Now what doesn't work is the setup of the sensor (initialization). For the sake of reference, I made two identical codes, one for the Pi Pico that has a problem, the other one for the RPi 4. The same sensor is used, and the code is completely identical. 

Lets start with the RPi 4 code that works completely as excepted:

```
import smbus
import time


MCP9600 = 0x66

TIntregval = 0x02 #reg int temp 
TKregval = 0x00   #reg k temp

deviceConfig = 0x06 #config adc
adc12bit = 0xE0 # 12-bit adc 
adc16bit = 0x20 # 16-bit adc 


sentypeconfig = 0x05 #select thermocouple 
selectKtype = 0x00   # select k type no filter
selectKtype_mid = 0x04   # select k type mid filter


i2c = smbus.SMBus(1)

i2c.write_byte_data(MCP9600, deviceConfig, adc16bit)  #this part of the code is running as expected 

i2c.write_byte_data(MCP9600, sentypeconfig, selectKtype_mid) # #this part of the code is running as expected 



while True: 
    valueI = i2c.read_i2c_block_data(MCP9600, TIntregval, 2)
    print(valueI)

    if (valueI[0] & 0b10000000) == 0b10000000:
        temperatureI = (valueI[0] * 16.0 + valueI[1] /16.0) - 4096.0
    else:
        temperatureI = (valueI[0] * 16.0 + valueI[1] /16.0)
            
    print(temperatureI)


    valueK = i2c.read_i2c_block_data(MCP9600, TKregval, 2)
    print(valueK)

    if (valueK[0] & 0b10000000) == 0b10000000:
        temperatureK = (valueK[0] * 16.0 + valueK[1] /16.0) - 4096.0
    else:
        temperatureK = (valueK[0] * 16.0 + valueK[1] /16.0)
            
    print(temperatureK)
    
    time.sleep(0.5)

```

This is the output of the code captured by a cheap logic analyser and everything is just fine: 

![Rpi4](https://github.com/micropython/micropython/assets/101253191/998e6cf7-2f6c-4b8e-87e3-38edea8ec7d8)

Now for the Pi Pico code:

```
import machine
import time

i2c1 = machine.I2C(1, sda=machine.Pin(14), scl=machine.Pin(15), freq=50000) #Error 5 change CLK speed
                                                            #or wrog value CLK

MCP9600 = 0x66

TIntregval = 0x02 #reg int temp 
TKregval = 0x00   #reg k temp

deviceConfig = 0x06 #config adc
adc12bit = 0xE0 # 12-bit adc 
adc16bit = 0x20 # 16-bit adc 


sentypeconfig = 0x05 #select thermocouple 
selectKtype = 0x00   # select k type no filter
selectKtype_mid = 0x04   # select k type mid filter

i2c1.writeto_mem(MCP9600, deviceConfig, bytearray(adc16bit))   #this part of the code that has a problem 
i2c1.writeto_mem(MCP9600, sentypeconfig, bytearray(selectKtype_mid))  #this part of the code that has a problem 

valueK = [0x00,0x00]


while True:
    
    
    valueI = i2c1.readfrom_mem(MCP9600, TIntregval, 2)
    print(valueI)

    if (valueI[0] & 0b10000000) == 0b10000000:
        temperatureI = (valueI[0] * 16.0 + valueI[1] /16.0) - 4096.0
    else:
        temperatureI = (valueI[0] * 16.0 + valueI[1] /16.0)
            
    print(temperatureI)


    valueK = i2c1.readfrom_mem(MCP9600, TKregval, 2)
    print(valueK)

    if (valueK[0] & 0b10000000) == 0b10000000:
        temperatureK = (valueK[0] * 16.0 + valueK[1] /16.0) - 4096.0
    else:
        temperatureK = (valueK[0] * 16.0 + valueK[1] /16.0)
            
    print(temperatureK)
    
    time.sleep(0.5)
```
The output from the Pi Pico: 

![Pipicoread](https://github.com/micropython/micropython/assets/101253191/d457ce4f-3677-4993-b917-7bc856aa4edb)

As you can see in the two first write to functions there is a lot of unexpected garbage data (0x00) and the values that I want to write to the registers are gone (0x20 and 0x04).  This is where I configure the sensor. 

The code is quite literarily copy and paste. With the exception of the Micropython functions for I2C (i2c1.writeto_mem, i2c1.readfrom_mem).

The physical connection is also identical on a bredboard. I use the latest version of Micropython.  

I don't really understand what the problem is. Hope someone here can share some light on the situation. 

Thank you so much for your time reading all of this!

**Comment by robert-hh** [comment, marked as answer]
bytearray(adc16bit) creates an array with a length of 32 bytes with a value of  0, since adc16bit has a value of 32. Is that intentional? Or did you want to write bytearray([adc16bit])?

**Reply by ajevticc** [reply]
Thank you @robert-hh the second case was what I wanted to do. Now I understand the difference. 

Thank you!
